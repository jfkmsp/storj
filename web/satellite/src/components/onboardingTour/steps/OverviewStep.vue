// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="overview-area">
        <div class="overview-area__header">
            <WelcomeLeft />
            <h1 class="overview-area__header__title" aria-roledescription="title">Welcome</h1>
            <WelcomeRight />
        </div>
        <p class="overview-area__subtitle">Let's get you started using Storj</p>
        <p class="overview-area__question">Do you want to use web browser or command-line interface?</p>
        <div class="overview-area__routes">
            <OverviewContainer
                class="overview-area__routes__left-cont"
                is-web="true"
                title="Web browser"
                info="Start uploading files in the browser and instantly see how your data gets distributed over the Storj network around the world."
                button-label="Continue in web"
                :on-click="onUploadInBrowserClick"
                :is-disabled="isLoading"
            />
            <OverviewContainer
                title="Command line"
                info="The Uplink CLI is a command-line interface which allows you to upload and download files from the network, manage permissions and sharing."
                button-label="Continue in cli"
                :on-click="onUplinkCLIClick"
                :is-disabled="isLoading"
            />
        </div>
        <router-link
            class="overview-area__skip-button"
            :to="projectDashboardPath"
        >
            Skip and go directly to dashboard
        </router-link>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import OverviewContainer from '@/components/onboardingTour/steps/common/OverviewContainer.vue';
import WelcomeLeft from '@/../static/images/onboardingTour/welcome-left.svg';
import WelcomeRight from '@/../static/images/onboardingTour/welcome-right.svg';

import { AnalyticsHttpApi } from '@/api/analytics';
import { RouteConfig } from '@/router';
import { AnalyticsEvent } from '@/utils/constants/analyticsEventNames';

// @vue/component
@Component({
    components: {
        OverviewContainer,
        WelcomeLeft,
        WelcomeRight
    },
})
export default class OverviewStep extends Vue {
    public isLoading = false;
    public projectDashboardPath = RouteConfig.ProjectDashboard.path;

    private readonly analytics: AnalyticsHttpApi = new AnalyticsHttpApi();

    /**
     * Holds button click logic.
     * Redirects to next step (creating access grant).
     */
    public async onUplinkCLIClick(): Promise<void> {
        if (this.isLoading) return;

        this.isLoading = true;

        await this.analytics.linkEventTriggered(AnalyticsEvent.PATH_SELECTED, 'CLI');
        await this.$router.push(RouteConfig.OnboardingTour.with(RouteConfig.CLIStep).with(RouteConfig.APIKey).path);

        this.isLoading = false;
    }

    /**
     * Redirects to objects page.
     */
    public async onUploadInBrowserClick(): Promise<void> {
        if (this.isLoading) return;

        this.isLoading = true;

        await this.analytics.linkEventTriggered(AnalyticsEvent.PATH_SELECTED, 'Continue in Browser');
        await this.$router.push(RouteConfig.Buckets.path).catch(() => {return; });

        this.isLoading = false;
    }
}
</script>

<style scoped lang="scss">
    .overview-area {
        display: flex;
        flex-direction: column;
        align-items: center;
        font-family: 'font_regular', sans-serif;

        &__header {
            display: flex;

            &__title {
                font-family: 'font_bold', sans-serif;
                font-size: 48px;
                line-height: 48px;
                letter-spacing: 1px;
                color: #14142b;
                margin: 0 20px 20px;
            }
        }

        &__subtitle,
        &__question {
            font-family: 'font_medium', sans-serif;
            font-size: 16px;
            line-height: 32px;
            text-align: center;
            letter-spacing: 0.75px;
            color: #14142a;
        }

        &__question {
            font-family: 'font_regular', sans-serif;
        }

        &__routes {
            margin-top: 70px;
            display: flex;
            align-items: center;

            &__left-cont {
                margin-right: 45px;
            }
        }

        &__skip-button {
            margin-top: 50px;
            color: #b7c1ca;
            cursor: pointer;
            text-decoration: underline !important;

            &:hover {
                text-decoration: underline;
            }
        }
    }
</style>
