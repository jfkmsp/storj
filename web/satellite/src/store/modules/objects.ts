// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    S3Client,
    Bucket,
    ListBucketsCommand,
    CreateBucketCommand,
    DeleteBucketCommand
} from '@aws-sdk/client-s3';

import { EdgeCredentials } from '@/types/accessGrants';
import { APP_STATE_ACTIONS } from '@/utils/constants/actionNames';
import { FilesState } from '@/store/modules/files';
import {StoreModule} from "@/types/store";
import {SignatureV4} from "@aws-sdk/signature-v4";

export const OBJECTS_ACTIONS = {
    CLEAR: 'clearObjects',
    SET_GATEWAY_CREDENTIALS: 'setGatewayCredentials',
    SET_API_KEY: 'setApiKey',
    SET_S3_CLIENT: 'setS3Client',
    SET_PASSPHRASE: 'setPassphrase',
    SET_FILE_COMPONENT_BUCKET_NAME: 'setFileComponentBucketName',
    FETCH_BUCKETS: 'fetchBuckets',
    CREATE_BUCKET: 'createBucket',
    CREATE_DEMO_BUCKET: 'createDemoBucket',
    DELETE_BUCKET: 'deleteBucket',
    CHECK_ONGOING_UPLOADS: 'checkOngoingUploads',
};

export const OBJECTS_MUTATIONS = {
    SET_GATEWAY_CREDENTIALS: 'setGatewayCredentials',
    SET_API_KEY: 'setApiKey',
    CLEAR: 'clearObjects',
    SET_S3_CLIENT: 'setS3Client',
    SET_BUCKETS: 'setBuckets',
    SET_FILE_COMPONENT_BUCKET_NAME: 'setFileComponentBucketName',
    SET_PASSPHRASE: 'setPassphrase',
    SET_LEAVE_ROUTE: 'setLeaveRoute',
};

export const DEMO_BUCKET_NAME = 'demo-bucket';

const {
    CLEAR,
    SET_API_KEY,
    SET_GATEWAY_CREDENTIALS,
    SET_S3_CLIENT,
    SET_BUCKETS,
    SET_PASSPHRASE,
    SET_FILE_COMPONENT_BUCKET_NAME,
    SET_LEAVE_ROUTE,
} = OBJECTS_MUTATIONS;

export class ObjectsState {
    public apiKey = '';
    public gatewayCredentials: EdgeCredentials = new EdgeCredentials();
    public s3Client: S3Client = new S3Client({
        forcePathStyle: true,
        signerConstructor: SignatureV4,
        // TODO: httpOptions: { timeout: 0 },
    });
    public bucketsList: Bucket[] = [];
    public passphrase = '';
    public fileComponentBucketName = '';
    public leaveRoute = '';
}

interface ObjectsContext {
    state: ObjectsState
    commit: (string, ...unknown) => void
    dispatch: (string, ...unknown) => Promise<any> // eslint-disable-line @typescript-eslint/no-explicit-any
    rootState: {
        files: FilesState
    }
}

/**
 * Creates objects module with all dependencies.
 */
export function makeObjectsModule(): StoreModule<ObjectsState, ObjectsContext> {
    return {
        state: new ObjectsState(),
        mutations: {
            [SET_API_KEY](state: ObjectsState, apiKey: string) {
                state.apiKey = apiKey;
            },
            [SET_GATEWAY_CREDENTIALS](state: ObjectsState, credentials: EdgeCredentials) {
                state.gatewayCredentials = credentials;
            },
            [SET_S3_CLIENT](state: ObjectsState) {
                state.s3Client = new S3Client({
                    credentials: {
                        accessKeyId: state.gatewayCredentials.accessKeyId,
                        secretAccessKey: state.gatewayCredentials.secretKey,
                    },
                    endpoint: state.gatewayCredentials.endpoint,
                    forcePathStyle: true,
                    signerConstructor: SignatureV4,
                    // TODO: httpOptions: { timeout: 0 },
                });
            },
            [SET_BUCKETS](state: ObjectsState, buckets: Bucket[]) {
                state.bucketsList = buckets;
            },
            [SET_PASSPHRASE](state: ObjectsState, passphrase: string) {
                state.passphrase = passphrase;
            },
            [SET_FILE_COMPONENT_BUCKET_NAME](state: ObjectsState, bucketName: string) {
                state.fileComponentBucketName = bucketName;
            },
            [SET_LEAVE_ROUTE](state: ObjectsState, leaveRoute: string) {
                state.leaveRoute = leaveRoute;
            },
            [CLEAR](state: ObjectsState) {
                state.apiKey = '';
                state.passphrase = '';
                state.gatewayCredentials = new EdgeCredentials();
                state.s3Client = new S3Client({
                    forcePathStyle: true,
                    signerConstructor: SignatureV4,
                    // TODO: httpOptions: { timeout: 0 },
                });
                state.bucketsList = [];
                state.fileComponentBucketName = '';
            },
        },
        actions: {
            setApiKey: function({commit}: ObjectsContext, apiKey: string): void {
                commit(SET_API_KEY, apiKey);
            },
            setGatewayCredentials: function({commit}: ObjectsContext, credentials: EdgeCredentials): void {
                commit(SET_GATEWAY_CREDENTIALS, credentials);
            },
            setS3Client: function({commit}: ObjectsContext): void {
                commit(SET_S3_CLIENT);
            },
            setPassphrase: function({commit}: ObjectsContext, passphrase: string): void {
                commit(SET_PASSPHRASE, passphrase);
            },
            setFileComponentBucketName: function({commit}: ObjectsContext, bucketName: string): void {
                commit(SET_FILE_COMPONENT_BUCKET_NAME, bucketName);
            },
            fetchBuckets: async function(ctx): Promise<void> {
                const result = await ctx.state.s3Client.send(new ListBucketsCommand({}));
                ctx.commit(SET_BUCKETS, result.Buckets);
            },
            createBucket: async function(ctx, name: string): Promise<void> {
                await ctx.state.s3Client.send(new CreateBucketCommand({
                    Bucket: name,
                }))
            },
            createDemoBucket: async function(ctx): Promise<void> {
                await ctx.state.s3Client.send(new CreateBucketCommand({
                    Bucket: DEMO_BUCKET_NAME,
                }))
            },
            deleteBucket: async function(ctx, name: string): Promise<void> {
                await ctx.state.s3Client.send(
                    new DeleteBucketCommand({
                        Bucket: name,
                    }));
            },
            clearObjects: function({commit}: ObjectsContext): void {
                commit(CLEAR);
            },
            checkOngoingUploads: function({commit, dispatch, rootState}: ObjectsContext, leaveRoute: string): boolean {
                if (!rootState.files.uploading.length) {
                    return false;
                }

                commit(SET_LEAVE_ROUTE, leaveRoute);
                dispatch(APP_STATE_ACTIONS.TOGGLE_UPLOAD_CANCEL_POPUP, null, {root: true});

                return true;
            },
        },
    };
}
