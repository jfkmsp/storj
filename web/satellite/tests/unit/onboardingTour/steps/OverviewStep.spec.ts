// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import { router } from '@/router';
import { appStateModule } from '@/store/modules/appState';

import OverviewStep from '@/components/onboardingTour/steps/OverviewStep.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store({ modules: { appStateModule } });

describe('OverviewStep.vue', (): void => {
    it('renders correctly', (): void => {
        const wrapper = shallowMount(OverviewStep, {
            localVue,
            router,
            store,
        });

        expect(wrapper).toMatchSnapshot();
    });
});
