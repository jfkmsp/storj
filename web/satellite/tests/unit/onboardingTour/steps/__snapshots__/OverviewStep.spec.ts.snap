// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewStep.vue renders correctly 1`] = `
<div class="overview-area">
  <div class="overview-area__header"><svg></svg>
    <h1 aria-roledescription="title" class="overview-area__header__title">Welcome</h1> <svg></svg>
  </div>
  <p class="overview-area__subtitle">Let's get you started using Storj</p>
  <p class="overview-area__question">Do you want to use web browser or command-line interface?</p>
  <div class="overview-area__routes">
    <div class="overview-container overview-area__routes__left-cont"><svg alt="web" class="overview-container__img"></svg>
      <h2 class="overview-container__title">Web browser</h2>
      <p aria-roledescription="server-side-encryption-title" class="overview-container__enc">Server-side encrypted</p>
      <p class="overview-container__info">Start uploading files in the browser and instantly see how your data gets distributed over the Storj network around the world.</p>
      <p class="overview-container__encryption-container">
        By using the web browser you are opting in to
        <a href="https://docs.storj.io/concepts/encryption-key/design-decision-server-side-encryption" target="_blank" rel="noopener noreferrer" aria-roledescription="server-side-encryption-link" class="overview-container__encryption-container__link">server-side encryption</a>.
      </p>
      <div class="container" style="width: 100%; height: 64px; border-radius: 62px; font-size: 16px;"> <span class="label uppercase">Continue in web</span></div>
    </div>
    <div class="overview-container"><svg alt="cli" class="overview-container__img"></svg>
      <h2 class="overview-container__title">Command line</h2>
      <p aria-roledescription="end-to-end-encryption-title" class="overview-container__enc">End-to-end encrypted</p>
      <p class="overview-container__info">The Uplink CLI is a command-line interface which allows you to upload and download files from the network, manage permissions and sharing.</p>
      <p class="overview-container__encryption-container">
        The Uplink CLI uses
        <a href="https://docs.storj.io/concepts/encryption-key/design-decision-end-to-end-encryption" target="_blank" rel="noopener noreferrer" aria-roledescription="end-to-end-encryption-link" class="overview-container__encryption-container__link">end-to-end</a> encryption for object data, metadata and path data.
      </p>
      <div class="container" style="width: 100%; height: 64px; border-radius: 62px; font-size: 16px;"> <span class="label uppercase">Continue in cli</span></div>
    </div>
  </div> <a href="/project-dashboard" class="overview-area__skip-button">
    Skip and go directly to dashboard
  </a>
</div>
`;
